dxr/plugins/python/indexers.py

```diff
@@ -122,12 +124,84 @@ def file_to_index(self, path, contents):
                                self.derived_classes, self.names)
 
 
-class FileToIndex(FileToIndexBase):
-    needle_types = {
-        ast.ClassDef: 'py_type',
-        ast.FunctionDef: 'py_function',
-    }
+class IndexingNodeVisitor(ast.NodeVisitor):
+    """Node visitor that walks through the nodes in an abstract syntax
+    tree and finds interesting things to index.
+
+    """
+
+    def __init__(self, file_to_index):
+        self.file_to_index = file_to_index
+        self.function_call_stack = []
+        self._needles = None
+
+    def get_needles(self, node):
+        """Walk through the given AST and return needles to index from
+        the tree.
+
+        """
+        self._needles = []
+        self.visit(node)
+        return self._needles
+
+    def visit_FunctionDef(self, node):
+        # Index the function itself for the function: filter.
+        start, end = self.file_to_index.get_node_start_end(node)
+        self.yield_needle('py_function', node.name, start, end)
+
+        # Index function calls within this function for the callers: and
+        # called-by filters.
+        self.function_call_stack.append([])
+        self.generic_visit(node)
```

Comment:
This append/visit/pop thing is elegant and charming. :-) You get to collect the results of visit_Call without having to rewrite the whole visitation framework to preserve its results. Nice work.